### PR TITLE
subscriber: For registry-backed layers, delay deletion of span until all layers have processed the event

### LINKED ISSUE #429: subscriber: Enable `Layer`s to read a recently-closed span
(This is followup from #420)

`Layer::on_close` should be able to read the _just_ closed span(s). At the moment, it's not possible to see a fully-completed span & export span's data, as the registry is not guaranteed to have the just-closed span in question.

### PR DESCRIPTION
**Pull request by davidbarsky**
This PR implements the second option ("The One With The `downcast_ref`") in @hawkw's [comment](https://github.com/tokio-rs/tracing/issues/429#issuecomment-554077428). The benchmarks of this change, relative to the master branch, are available [here](https://gist.github.com/davidbarsky/3620247e6cf74e77e1d5d4c24014c6d1). 

Edit: The same benchmarks running on a significantly less noisy c3.8xlarge: https://gist.github.com/davidbarsky/111814dcc55e89580e5ee283b9900bbf.

Resolves #429

### CODE REVIEW

**Comment by davidbarsky** [comment]
The same benchmarks running on a _significantly_ less noisy c3.8xlarge: https://gist.github.com/davidbarsky/111814dcc55e89580e5ee283b9900bbf.

I first ran benchmarks on the `master` branch and then on the `david/downcast_ref-registry` branch.

File: tracing-subscriber/src/layer.rs [hunk 1]
```diff
@@ -494,6 +494,12 @@ where
     }
 
     fn try_close(&self, id: span::Id) -> bool {
+        let subscriber = &self.inner as &dyn Subscriber;
+        let _guard = match subscriber.downcast_ref::<Registry>() {
+            Some(registry) => Some(registry.ref_guard(id.clone())),
+            None => None,
+        };
```
Comment (hawkw) [inline tracing-subscriber/src/layer.rs]: TIOLI: I think this could be 
```rust
let _guard = subscriber.downcast_ref::<Registry>()
    .and_then(|registry| registry.ref_guard(id.clone());
```
which could be a little more concise

Comment (hawkw) [inline tracing-subscriber/src/layer.rs]: also, this will need to be flagged with `#[cfg(feature = "registry")]; i don't think this code compiles with that feature disabled. so, something like:
```rust
#[cfg(feature = "registry")]
let subscriber = &self.inner as &dyn Subscriber;
#[cfg(feature = "registry")]
let _guard = ...;
```

File: tracing-subscriber/src/registry/sharded.rs [hunk 2]
```diff
@@ -89,6 +89,24 @@ fn id_to_idx(id: &Id) -> usize {
     id.into_u64() as usize - 1
 }
 
+pub(crate) struct LayerGuard<'a> {
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: I don't think this name really makes it clear what this thing is for. What's a "LayerGuard"? Why "Layer"? I might call it something like `CloseGuard`, and maybe add a comment?

File: tracing-subscriber/src/registry/sharded.rs [hunk 3]
```diff
@@ -89,6 +89,24 @@ fn id_to_idx(id: &Id) -> usize {
...
+        CLOSE_COUNT.with(|count| {
+            let c = count.get();
+            if c > 0 {
+                count.set(c - 1);
+            } else {
+                self.registry.spans.remove(id_to_idx(&self.id));
+            }
+        })
+    }
+}
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: super nitpicky, but i might move this lower in the file, after the `Registry` impls?

File: tracing-subscriber/src/registry/sharded.rs [hunk 4]
```diff
@@ -89,6 +89,24 @@ fn id_to_idx(id: &Id) -> usize {
...
 }
 
+pub(crate) struct LayerGuard<'a> {
+    id: Id,
+    registry: &'a Registry,
+}
+
+impl<'a> Drop for LayerGuard<'a> {
+    fn drop(&mut self) {
+        CLOSE_COUNT.with(|count| {
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: `LocalKey::with` [will panic](https://doc.rust-lang.org/std/thread/struct.LocalKey.html#panics) if the thread local has already been destroyed. Since this is a `Drop` impl, it can be called while unwinding. 

We should avoid a double panic by calling [`LocalKey::try_with`](https://doc.rust-lang.org/std/thread/struct.LocalKey.html#method.try_with) instead. If the `try_with` fails, we are panicking, so I think okay to do nothing in that case and allow the span to be "leaked", since the panic will nuke the span storage anyway.

File: tracing-subscriber/src/registry/sharded.rs [hunk 5]
```diff
@@ -97,9 +115,29 @@ impl Registry {
     fn get(&self, id: &Id) -> Option<Guard<'_, DataInner>> {
         self.spans.get(id_to_idx(id))
     }
+
+    pub(crate) fn ref_guard(&self, id: Id) -> LayerGuard<'_> {
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: I don't find this name super descriptive of what this method _means_ — while I understand what's going on here, I don't think this would be very clear to a new contributor reading this code for the first time. 

What if we changed this name to something that explained _why_ we call this, rather than what it returns? Maybe `start_close` or something? Of course, if you can think of a better name, that would be great.

A comment explaining what this does could be good as well.

File: tracing-subscriber/src/registry/sharded.rs [hunk 6]
```diff
@@ -97,9 +115,29 @@ impl Registry {
...
 
 thread_local! {
+    // `CLOSE_COUNT` is used to track how many layers have processed a
+    // close event. Once the `CLOSE_COUNT` is 0, the registry knows that
+    // is is safe to remove a span. It does so via the drop on
+    // `LayerGuard`.
+    //
+    // This behavior is needed to enable a Registry-backed Layer to
+    // access span data after the Layer has recieved the `on_close`
+    // callback.
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: this comment is good, but i am not sure if the thread local is the right place for it to be...I might move most of this to the method currently called `ref_guard`, and maybe leave a shorter comment here.

File: tracing-subscriber/src/registry/sharded.rs [hunk 7]
```diff
@@ -190,7 +228,6 @@ impl Subscriber for Registry {
             None if std::thread::panicking() => return false,
             None => panic!("tried to drop a ref to {:?}, but no such span exists!", id),
         };
-
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: unrelated whitespace change?

**Comment by hawkw** [comment]
>The benchmarks of this change, relative to the master branch, are available [here](https://gist.github.com/davidbarsky/3620247e6cf74e77e1d5d4c24014c6d1).

On Discord, we discussed how these benchmarks might potentially be inaccurate (perhaps due to a noisy benchmarking environment). Did you have a chance to run them again in a more isolated benchmark environment? If so, it would be great to see those results as well.

File: tracing-subscriber/src/layer.rs (hunk 1 above)
Comment (davidbarsky) [reply tracing-subscriber/src/layer.rs]: Good point, fixed.

File: tracing-subscriber/src/registry/sharded.rs (hunk 4 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: fixed, thanks!

**Comment by hawkw** [comment]
> The same benchmarks running on a _significantly_ less noisy c3.8xlarge: https://gist.github.com/davidbarsky/111814dcc55e89580e5ee283b9900bbf.
> 
> I first ran benchmarks on the `master` branch and then on the `david/downcast_ref-registry` branch.

Great, thank you — I must've missed this earlier. Seems like the performance difference is very slight, _probably_ not significant enough to justify the more complex unsafe solution. We can revisit this if it turns out to have an impact.

File: tracing-subscriber/src/registry/sharded.rs (hunk 2 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: I've opted for calling it `CloseGuard` & added a comment describing what this is intended for.

File: tracing-subscriber/src/registry/sharded.rs (hunk 3 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Moved next to DataInner's drop impl.

File: tracing-subscriber/src/registry/sharded.rs (hunk 5 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Added details and renamed to `start_close`. I couldn't think of anything better :)

File: tracing-subscriber/src/registry/sharded.rs (hunk 6 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Done!

File: tracing-subscriber/src/layer.rs (hunk 1 above)
Comment (davidbarsky) [reply tracing-subscriber/src/layer.rs]: done!

File: tracing-subscriber/src/registry/sharded.rs (hunk 7 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: fixed, thanks for noticing :)

**Comment by davidbarsky** [comment]
@hawkw I _think_ I've addressed all of the comments you had, including adding tests. I'd appreciate a re-review whenever you have the time!

File: tracing-subscriber/src/registry/sharded.rs [hunk 8]
```diff
@@ -270,3 +320,100 @@ impl<'a> SpanData<'a> for Data<'a> {
...
+    use super::Registry;
+    use crate::{layer::Context, registry::LookupSpan, Layer};
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use tracing::{self, subscriber::with_default};
+    use tracing_core::{
+        span::{Attributes, Id},
+        Subscriber,
+    };
+
+    struct NopLayer;
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: Nit: I wouldn't call this a "nop" layer; it makes an assertion...not a blocker.

File: tracing-subscriber/src/registry/sharded.rs [hunk 9]
```diff
@@ -270,3 +320,100 @@ impl<'a> SpanData<'a> for Data<'a> {
...
+            let span = tracing::debug_span!("span");
+            drop(span);
+        });
+    }
+
+    #[test]
+    fn multiple_layers_can_access_closed_span() {
+        let subscriber = NopLayer
+            .and_then(NopLayer)
+            .and_then(NopLayer2)
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: Nit/TIOLI: two separate layer types isn't _strictly_ necessary, you could just use a whole bunch of layers with the same type...not a big deal, and I see why we _might_ want to test that different layers of different types also see the right behavior, but there would be less boilerplate.

File: tracing-subscriber/src/registry/sharded.rs [hunk 10]
```diff
@@ -270,3 +320,100 @@ impl<'a> SpanData<'a> for Data<'a> {
...
+            S: Subscriber + for<'a> LookupSpan<'a>,
+        {
+            fn new_span(&self, _: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+                let span = ctx.span(id).expect("Missing span; this is a bug");
+                let mut extensions = span.extensions_mut();
+                extensions.insert(ClosingSpan);
+            }
+
+            fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+                assert!(&ctx.span(&id).is_some());
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: should we maybe also assert the extension is present here? It should be, though.

File: tracing-subscriber/src/registry/sharded.rs [hunk 11]
```diff
@@ -89,6 +89,17 @@ fn id_to_idx(id: &Id) -> usize {
     id.into_u64() as usize - 1
 }
 
+// CloseGuard is used to track how many Registry-backed Layers have
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: Nit/TIOLI: RustDoc comments usually don't start with the type names...

```suggestion
// A guard that tracks how many Registry-backed Layers have
```

File: tracing-subscriber/src/registry/sharded.rs [hunk 12]
```diff
@@ -89,6 +89,17 @@ fn id_to_idx(id: &Id) -> usize {
...
+// processed an `on_close` event. Once all Layers have processed this
+// event, the registry knows that is able to safely remove the span
+// tracked by `id`.
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: How does the guard help with this?

File: tracing-subscriber/src/registry/sharded.rs [hunk 13]
```diff
@@ -97,9 +108,33 @@ impl Registry {
     fn get(&self, id: &Id) -> Option<Guard<'_, DataInner>> {
         self.spans.get(id_to_idx(id))
     }
+
+    // `start_close` creates a guard which tracks how many layers have
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: nit:
```suggestion
    // Returns a guard which tracks how many layers have
```

File: tracing-subscriber/src/registry/sharded.rs [hunk 14]
```diff
@@ -97,9 +108,33 @@ impl Registry {
...
+    // processed a close event via the `CLOSE_COUNT` thread-local. Once
+    // the `CLOSE_COUNT` is 0, the registry knows that is is safe to
+    // remove a span. It does so via the Drop implementation on
+    // `CloseGuard`.
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: I guess this answers the "how does it work" question in the comment I left on the other doc comment?

File: tracing-subscriber/src/registry/sharded.rs [hunk 15]
```diff
@@ -89,6 +89,17 @@ fn id_to_idx(id: &Id) -> usize {
     id.into_u64() as usize - 1
 }
 
+// CloseGuard is used to track how many Registry-backed Layers have
+// processed an `on_close` event. Once all Layers have processed this
+// event, the registry knows that is able to safely remove the span
+// tracked by `id`.
+//
+// For additional details, see the comment on `Registry::start_close`.
```
Comment (hawkw) [inline tracing-subscriber/src/registry/sharded.rs]: Nit: even though these are private APIs, I think we should use three-slash doc comments for them...not a blocker.

File: tracing-subscriber/src/registry/sharded.rs (hunk 10 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: agreed, we should.

File: tracing-subscriber/src/registry/sharded.rs (hunk 8 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: called it "AssertionLayer" and "AssertionLayer2".

File: tracing-subscriber/src/registry/sharded.rs (hunk 9 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: hrm, fair point. They would just be called `n` times, where `n` is the amount of times we compose those layers.

File: tracing-subscriber/src/registry/sharded.rs (hunk 10 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Added.

File: tracing-subscriber/src/registry/sharded.rs (hunk 8 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Actually, removed `AssertionLayer2` and composed it twice. Works well!

File: tracing-subscriber/src/registry/sharded.rs (hunk 15 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Added.

File: tracing-subscriber/src/registry/sharded.rs (hunk 14 above)
Comment (davidbarsky) [reply tracing-subscriber/src/registry/sharded.rs]: Yep, I did some more writing.

**Comment by hawkw** [comment]
the test failures on nightly (https://github.com/tokio-rs/tracing/pull/432/checks?check_run_id=306313856) seem unrelated, but it would be good if they could be addressed

**Comment by davidbarsky** [comment]
@hawkw Sorry—just saw your comment about nightly failures. I'll address that in a separate branch.
